src/mococrw/dilithium.h

```diff
@@ -0,0 +1,392 @@
+/*
+ * #%L
+ * %%
+ * Copyright (C) 2022 BMW Car IT GmbH
+ * %%
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ * #L%
+ */
+#pragma once
+
+#include <cstdint>
+#include <memory>
+#include <vector>
+#include "mococrw/asymmetric_crypto_ctx.h"
+#include "mococrw/key.h"
+
+namespace mococrw
+{
+/*
+ * The idea is to implement dilithium in a way that we can later alias the classes here to the ones
+ * in key.h and asymmetric_crypto_ctx.h with reduced efforts in code adaptions.
+ */
+
+/**
+ * This class is a replacement of EVP_PKEY as long as dilithium is not supported by the openssl
+ * version used by MoCOCrW. It stores the key as std::vector<uint8_t> and the key type.
+ * Addittionally some convenient functions are added.
+ */
+class DilithiumKeyImpl
+{
+public:
+    enum class DilithiumParameterSet : int { DILITHIUM2, DILITHIUM3, DILITHIUM5 };
+
+    DilithiumKeyImpl(const std::vector<uint8_t> key_data,
+                     DilithiumKeyImpl::DilithiumParameterSet paramSet,
+                     bool is_private_key)
+            : _key_data{std::move(key_data)}, _is_private_key(is_private_key), _paramSet(paramSet)
+    {
+        if (!hasValidKeySize()) {
+            throw MoCOCrWException("Key data does not match the expected size.");
+        }
+    }
+
+    /**
+     * @brief Parses a ASN.1 X509 public key object (see RFC5280)
+     *
+     * @param x509PubKey The public key object as defined in RFC5280
+     * @return A shared pointer to a dilithium implementation object
+     */
+    static std::shared_ptr<DilithiumKeyImpl> readPublicKeyFromDER(
+            const std::vector<uint8_t> &x509PubKey);
+
+    /**
+     * @brief Parses a ASN.1 PKCS#8 private key object (see RFC5208)
+     *
+     * @param x509PubKey The private key object as defined in RFC5208
+     * @return A shared pointer to a dilithium implementation object
+     */
+    static std::shared_ptr<DilithiumKeyImpl> readPrivateKeyFromDER(
+            const std::vector<uint8_t> &pkcs8PrivKey);
+
+    /**
+     * @returns The key type
+     */
+    AsymmetricKey::KeyTypes getType() const { return _keyType; };
+
+    /**
+     * @returns The size of the raw key data
+     */
+    uint getKeySize() const { return _key_data.size(); }
+
+    /**
+     * Returns the raw key data
+     *
+     * @returns The raw key data
+     */
+    const std::vector<uint8_t> getKeyData() const { return _key_data; };
```

Comment:
hm. Now that I thought about this. Maybe it would be better to return a ptr here. This currently creates wild copies of (potentially) private key data in memory that isn't later blanked out. It's not really relevant to our main use-cases but I don't think its good to leave it if we offer this functionality at all. Maybe we just offer a way to get the key data byte buffer directly for the implementation functions?